File: regions/C++/swap_regions/swapRegions.cpp
#include <iostream>
#include <opencv2/opencv.hpp>
#include <string>

#include "opencv2/core/types.hpp"
#include "opencv2/core/mat.hpp"

cv::Mat loadImage(std::string pathImage);

void showImage(std::string title, cv::Mat& image);

int main(int argc, char** argv) {

    cv::Mat image = loadImage(argv[1]);

    cv::Mat new_image;

    image.copyTo(new_image);


    cv::Rect A(0, 0, (int) image.rows / 2, (int) image.cols / 2);

    cv::Rect B((int) image.rows / 2, 0, (int) image.rows / 2, (int) image.cols / 2);

    cv::Rect C(0, (int) image.cols / 2, (int) image.rows / 2, (int) image.cols / 2);

    cv::Rect D((int) image.rows / 2, (int) image.cols / 2, (int) image.rows / 2, (int) image.cols / 2);

    image(A).copyTo(new_image(D));

    image(D).copyTo(new_image(A));

    image(B).copyTo(new_image(C));

    image(C).copyTo(new_image(B));


    showImage("puzzle biel", new_image);

    return 0;

}

void showImage(std::string title, cv::Mat& image) {
    cv::namedWindow(title, cv::WINDOW_NORMAL);

    cv::imshow(title, image);
    cv::waitKey();
}

cv::Mat loadImage(std::string pathImage) {
    cv::Mat image = cv::imread(pathImage, cv::IMREAD_GRAYSCALE);

    if (not image.data) {
        std::cout << "imagem não carregada" << std::endl;
    }

    return image;
}

File: regions/Python/swap_regions.py
import cv2
import sys
import numpy as np


def show_image(title: str, image: np.ndarray):
    cv2.namedWindow(title, cv2.WINDOW_NORMAL)
    cv2.imshow(title, image)
    cv2.waitKey(0)


def get_regions(image: np.ndarray) -> list:
    rows = image.shape[0]
    cols = image.shape[1]

    return [
        image[int(rows / 2):rows, int(cols / 2):cols],
        image[int(rows / 2):rows, 0:int(cols / 2)],
        image[0:int(rows / 2), int(cols / 2):cols],
        image[0:int(rows / 2), 0:int(cols / 2)]

    ]


def make_new_image(regions: list) -> np.ndarray:
    a = np.hstack(np.array(regions[0:2]))

    b = np.hstack(np.array(regions[2:4]))

    return np.vstack((a, b))


def main():
    image = cv2.imread(sys.argv[1], cv2.IMREAD_GRAYSCALE)

    regions = get_regions(image)

    new_image = make_new_image(regions)

    show_image("puzzle image", new_image)


if __name__ == '__main__':
    main()
